Social Login: Offer to create a new account if it doesn't exist

zaguiini opened the pull request:
#### Proposed Changes

Add the URL to the signup page in case the user faces the `unknown_user` message when trying to social login:

<img width="925" alt="image" src="https://user-images.githubusercontent.com/36432/182399750-df31b947-273e-40f1-afda-fcca1a992e01.png">

~~I tried passing the service as a query parameter to the signup URL and opening the pop-up on page load, but the browser blocks it. At least Brave did... Chrome also did but it's possible to allow it. Regardless, I'm committing the changes because the user might allow pop-ups by default.~~ Decided to remove the auto-popup behavior.

#### Testing Instructions

See "Testing Instructions" in https://github.com/Automattic/wp-calypso/pull/64957 to set up a working wpcalypso.wordpress.com environment on your local machine.

Try to social login using an account that is not registered in WPCOM and you should see the above message with a link to the signup page. It should work with both Google and Apple.

Clicking on the link should dispatch the `calypso_login_sign_up_link_click` Tracks event.

Fixes https://github.com/Automattic/wp-calypso/issues/65651

zaguiini committed:
Add sign up link to the unknown_user message

zaguiini commented:
```diff
diff --git a/client/blocks/login/error-notice.jsx b/client/blocks/login/error-notice.jsx
index 601b704358..72742c7309 100644
--- a/client/blocks/login/error-notice.jsx
+++ b/client/blocks/login/error-notice.jsx
@@ -4,11 +4,13 @@ import PropTypes from 'prop-types';
 import { Component } from 'react';
 import { connect } from 'react-redux';
 import Notice from 'calypso/components/notice';
+import { addQueryArgs } from 'calypso/lib/url';
 import {
 	getRequestError,
 	getTwoFactorAuthRequestError,
 	getCreateSocialAccountError,
 	getRequestSocialAccountError,
+	getSocialAccountLinkService,
 } from 'calypso/state/login/selectors';
 
 class ErrorNotice extends Component {
@@ -71,13 +73,23 @@ class ErrorNotice extends Component {
 
 		let message = error.message;
 
-		if ( error.code === 'unknown_user' && this.props.signupUrl ) {
+		const { signupUrl, socialAccountLinkService } = this.props;
+
+		if ( error.code === 'unknown_user' && signupUrl ) {
 			message = this.props.translate(
 				"Hmm, we can't find a WordPress.com account for that social login. Please double check your information and try again. " +
 					'Alternatively, you can try {{a}}creating a new account{{/a}}.',
 				{
 					components: {
-						a: <a href={ this.props.signupUrl } />,
+						a: (
+							<a
+								href={
+									socialAccountLinkService
+										? addQueryArgs( { launch_service: socialAccountLinkService }, signupUrl )
+										: signupUrl
+								}
+							/>
+						),
 					},
 				}
 			);
@@ -115,4 +127,5 @@ export default connect( ( state ) => ( {
 	requestAccountError: getRequestSocialAccountError( state ),
 	requestError: getRequestError( state ),
 	twoFactorAuthRequestError: getTwoFactorAuthRequestError( state ),
+	socialAccountLinkService: getSocialAccountLinkService( state ),
 } ) )( localize( ErrorNotice ) );
diff --git a/client/components/social-buttons/google.js b/client/components/social-buttons/google.js
index 0f3d8872f1..1af87179dc 100644
--- a/client/components/social-buttons/google.js
+++ b/client/components/social-buttons/google.js
@@ -101,6 +101,10 @@ class GoogleLoginButton extends Component {
 		} );
 
 		this.setState( { isDisabled: false } );
+
+		if ( this.props.launchServiceOnLoad ) {
+			this.client.requestCode();
+		}
 	}
 
 	async loadGoogleIdentityServicesAPI() {
@@ -352,6 +356,7 @@ export default connect(
 	( state ) => ( {
 		isFormDisabled: isFormDisabled( state ),
 		authCodeFromRedirect: getInitialQueryArguments( state ).code,
+		launchServiceOnLoad: getInitialQueryArguments( state ).launch_service === 'google',
 	} ),
 	{
 		recordTracksEvent,
diff --git a/client/state/login/reducer.js b/client/state/login/reducer.js
index b33f3d255c..1a2aa45334 100644
--- a/client/state/login/reducer.js
+++ b/client/state/login/reducer.js
@@ -370,10 +370,10 @@ export const socialAccount = ( state = { createError: null }, action ) => {
 	return state;
 };
 
-const userExistsErrorHandler = ( state, { error, authInfo } ) => {
-	if ( error.code === 'user_exists' ) {
+const socialResponseErrorHandler = ( state, { error, authInfo } ) => {
+	if ( error.code === 'user_exists' || error.code === 'unknown_user' ) {
 		return {
-			isLinking: true,
+			isLinking: error.code === 'user_exists',
 			email: error.email,
 			authInfo,
 		};
@@ -385,9 +385,9 @@ const userExistsErrorHandler = ( state, { error, authInfo } ) => {
 export const socialAccountLink = ( state = { isLinking: false }, action ) => {
 	switch ( action.type ) {
 		case SOCIAL_CREATE_ACCOUNT_REQUEST_FAILURE:
-			return userExistsErrorHandler( state, action );
+			return socialResponseErrorHandler( state, action );
 		case SOCIAL_LOGIN_REQUEST_FAILURE:
-			return userExistsErrorHandler( state, action );
+			return socialResponseErrorHandler( state, action );
 		case SOCIAL_CONNECT_ACCOUNT_REQUEST_SUCCESS:
 			return { isLinking: false };
 		case CURRENT_USER_RECEIVE:

```

zaguiini committed:
Try to start the Google flow if launch_service query parameter is added

zaguiini committed:
Try to start the Apple flow if launch_service query parameter is added

zaguiini commented on client/components/social-buttons/google.js:
```diff
@@ -101,6 +101,10 @@ class GoogleLoginButton extends Component {
 
 		this.setState( { isDisabled: false } );
+
+		if ( this.props.launchServiceOnLoad ) {
```
Should we log another event here, something like `calypso_social_button_launched_on_load`?

zaguiini committed:
Log sign up link click event

zaguiini commented on client/blocks/login/error-notice.jsx:
```diff
@@ -53,6 +57,12 @@ class ErrorNotice extends Component {
 
+	registerSignupUrlClickEvent = () => {
+		this.props.recordTracksEvent( 'calypso_login_sign_up_link_click', {
+			reason: 'unknown_user',
```
Is this event prop valuable? The idea is to identify the origin. Maybe something like `origin: 'unknown_user_login_notice'` would be better...

vykes-mac replied:
if this `unknow_user` error can occur at other places where `calypso_login_sign_up_link_click` track is recorded then it might be more valuable to give additional context to the reason for clarity. If not it should be good as is.

zaguiini commented:
Not really sure why E2E tests are failing here, but they're failing consistently.

zaguiini commented on client/blocks/login/error-notice.jsx:
```diff
@@ -70,6 +80,29 @@ class ErrorNotice extends Component {
+
+		if ( error.code === 'unknown_user' && signupUrl ) {
+			message = this.props.translate(
+				"Hmm, we can't find a WordPress.com account for that social login. Please double check your information and try again. " +
+					'Alternatively, you can try {{a}}creating a new account{{/a}}.',
```
Confirm copy.

zaguiini replied:
Confirm copy?

zaguiini commented on client/components/social-buttons/google.js:
```diff
@@ -101,6 +101,10 @@ class GoogleLoginButton extends Component {
 		this.setState( { isDisabled: false } );
+
+		if ( this.props.launchServiceOnLoad ) {
+			this.client.requestCode();
```
We can probably wrap this with `try/catch` so it doesn't register an error in the console. Additionally, an event (`calypso_social_button_launched_on_load_failure/success`) could be of some use.

danielbachhuber commented:
> I tried passing the service as a query parameter to the signup URL and opening the pop-up on page load, but the browser blocks it. At least Brave did... Chrome also did but it's possible to allow it. Regardless, I'm committing the changes because the user might allow pop-ups by default.

@vindl Do you have an opinion on this behavior? I'm inclined to remove it, if it will mostly only ever error.

vykes-mac commented:
> > I tried passing the service as a query parameter to the signup URL and opening the pop-up on page load, but the browser blocks it. At least Brave did... Chrome also did but it's possible to allow it. Regardless, I'm committing the changes because the user might allow pop-ups by default.
> 
> @vindl Do you have an opinion on this behavior? I'm inclined to remove it, if it will mostly only ever error.

I'm with you on this one @danielbachhuber. Probably could get around the pop-up issue buy simulating clicking the social button (finding element on the page and clicking it) but that might feels hacky also. Might be best to just redirect the user to the sign up page and let them continue with whatever social log in from there.

danielbachhuber committed:
Revert "Try to start the Apple flow if launch_service query parameter is added"

This reverts commit 82cfce91dbb18efa2ee8af6389617dbdad71b102.

danielbachhuber committed:
Manually revert 5532501c42629a5c34dd1d76679fc7677250f67f

danielbachhuber committed:
Reuse the existing string + more assertive statement for the second

danielbachhuber committed:
Remove unnecessary query arg; more precise Tracks event name

danielbachhuber commented on client/blocks/login/error-notice.jsx:
```diff
@@ -70,6 +72,27 @@ class ErrorNotice extends Component {
+
+		if ( error.code === 'unknown_user' && signupUrl ) {
+			message = this.props.translate(
+				"Hmm, we can't find a WordPress.com account for that social login. Please double check your information and try again." +
```
Would it be better to reuse the string that comes back directly from the API?

<img width="1411" alt="image" src="https://user-images.githubusercontent.com/36432/180864938-3c7206fc-711d-41e9-969c-3f30c746afc1.png">

vindl replied:
I think that's fine, as long as it's properly translated. We could move part of the existing message to a comment, clarifying what this state is all about.

danielbachhuber replied:
Added a clarifying comment with c97a7cb6182c5c8d98214767b64fca6193ec00f0

danielbachhuber committed:
Use `unknown_user` for greater clarity

danielbachhuber commented on client/blocks/login/error-notice.jsx:
```diff
@@ -70,6 +72,27 @@ class ErrorNotice extends Component {
+							<a
+								href={ signupUrl }
+								onClick={ this.props.recordTracksEvent(
+									'calypso_login_social_unknown_user_signup_link_click'
```
@vindl Are you fine with `calypso_login_social_unknown_user_signup_link_click` for this context?

vindl replied:
Yes. 👍

danielbachhuber replied:
Added to Tracks with 1005-gh-Automattic/tracks-events-registration

danielbachhuber commented on client/blocks/login/index.jsx:
```diff
@@ -197,6 +201,24 @@ class Login extends Component {
+			pathname,
+			isGutenboarding
+		);
+	};
```
Is this the best way to produce a signup URL?

p-jackson replied:
Yep this is the helper to use. There's quite a bit that goes into calculating the signup URL, and the user could have landed on the login page from lots of different places, so it's good to maintain all that context when we send them to the signup page.

p-jackson replied:
I think it's a little confusing that this `<Login>` block both accepts a `signupUrl` prop, and also has the ability to generate its own signup URL using this `getSignupUrl` method. Should this method be used everywhere the `signupUrl` prop is used? so it's more like providing a default value for the prop?

Another option might be to remove all this `getSignupUrl` logic from the `<Login>` component. Since all it ever does with this logic is just pass it on to the `<ErrorNotice>`, maybe the logic should just be moved inside `<ErrorNotice>` instead?

danielbachhuber replied:
> Yep this is the helper to use. There's quite a bit that goes into calculating the signup URL, and the user could have landed on the login page from lots of different places, so it's good to maintain all that context when we send them to the signup page

Great, thanks for clarifying.

> I think it's a little confusing that this `<Login>` block both accepts a `signupUrl` prop, and also has the ability to generate its own signup URL using this `getSignupUrl` method. Should this method be used everywhere the `signupUrl` prop is used? so it's more like providing a default value for the prop?

I think it's odd too, and I'm not sure what the original intent was. However, I feel reasonably confident @zaguiini had a rationale for it.

I'm open to your opinion on what should be done instead.

danielbachhuber replied:
@p-jackson Also, when testing in my local environment, `getSignupUrl` seems to produce `/start` instead of `/start/user`. Does this make sense to you? Anything we need to do to resolve?

p-jackson replied:
> I'm open to your opinion on what should be done instead.

I was thinking something like this https://github.com/Automattic/wp-calypso/pull/66148/files that moves the logic for generating the signup URL into the `<ErrorNotice>` since it seemed to be the only thing that used it.
But now that I've tried it I don't like it. It requires passing the `locale` and `isGutenboarding` props into the `<ErrorNotice>` component anyway, so I think it's just as tidy without my changes.

> `getSignupUrl` seems to produce `/start` instead of `/start/user`

Yep that's good. `/start` is the usual link we would use to start off the flow. The signup flow will decide whether they need to create an account or not and redirect them to the `user` step if necessary.

danielbachhuber commented on client/state/login/reducer.js:
```diff
@@ -370,10 +370,10 @@ export const socialAccount = ( state = { createError: null }, action ) => {
-const userExistsErrorHandler = ( state, { error, authInfo } ) => {
-	if ( error.code === 'user_exists' ) {
+const socialResponseErrorHandler = ( state, { error, authInfo } ) => {
+	if ( error.code === 'user_exists' || error.code === 'unknown_user' ) {
```
Need to look into this change.

danielbachhuber replied:
> Need to look into this change.

Ah, it was to set additional `store` state in 5532501c42629a5c34dd1d76679fc7677250f67f

Because I reverted that code in e5a8afd9c55b3b27bfc12588a3182ef9ed17979e, its safe to revert the remaining d2b05caa4b3c678807fe5c5538909d6d8f42ea90

danielbachhuber committed:
Clarify where this message is coming from

danielbachhuber committed:
Continue reverting unnecessary code from 5532501c42629a5c34dd1d76679fc7677250f67f

danielbachhuber committed:
Ensure the event isn't called immediately when the component is rendered

p-jackson reviewed (approved):
I've re-tested with both and Apple and Google account and this works as expected.
[I've rethought my earlier comment](https://github.com/Automattic/wp-calypso/pull/65909/files#r935007048) and I think this is good to go now. (if the CI manages to pass ... the failing test appears to be unrelated to these changes)

p-jackson commented:
I thought the failing e2e test was unrelated so I rebased in case the issue had already been fixed on `trunk`.

However it actually is an issue with this PR after all 🤦 
<img width="779" alt="CleanShot 2022-08-02 at 14 09 47@2x" src="https://user-images.githubusercontent.com/1500769/182276630-f2c67877-8a64-4893-9e66-e1222e25e84f.png">

Looks like this PR breaks SSR. So we might want to do something like what I suggested in #66148 after all. I'll make changes there and then you decide whether it makes sense @danielbachhuber

p-jackson commented:
@danielbachhuber I've updated #66148 so that it calculates the signup URL as late as possible. This makes it safe to access `window.location`. An alternative would be to find some server-compatible way to allow this error state to render. Maybe something much higher up the tree (that's closer to the controller code and so knows what the pathname is without accessing `window.location`) could calculate the signup URL and then pass it all the way down through the children.

p-jackson committed:
Move signurl logic into <ErrorNotice>

The calculation of the signup url is delayed as late as possible so that
it's only done on the client. Calculation of the url requires access to
`window.location` which can't be used on the server.

danielbachhuber committed:
Merge branch 'trunk' into improve/offer-to-create-account

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7462700

Thank you @zaguiini for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7462700).
